src/Featurizers/Components/StatisticsMetricsEstimator.h

```diff
@@ -0,0 +1,391 @@
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+
+// ----------------------------------------------------------------------
+// |
+// |  update functions
+// |
+// ----------------------------------------------------------------------
+
+template <typename T>
+void update_basic_statistics(T input, T& min, T& max, std::uint64_t& count, bool& first_element_flag) {
+    // first non-null element will set min and max
+    if (first_element_flag) {
+        min = input;
+        max = input;
+        ++count;
+        first_element_flag = false;
+        return;
+    }
+    if(input < min) {
+        min = input;
+    }
+    if(input > max) {
+        max = input;
+    }
+    // check if count will be out of bounds
+    if (std::numeric_limits<std::uint64_t>::max() == count) {
+        throw std::runtime_error("Overflow occured for count during calculating statistic metrics! Check your data!");
+    }
+    ++count;
+    return;
+}
+
+template <typename T>
+void update_standard_statistics(T input, long double& sum) {
+    if ((sum + static_cast<long double>(input) == sum) && (input != 0)) {
+        throw std::runtime_error("Input is so small comparing to sum that sum is the same after long double addition!");
+    }
+    sum += static_cast<long double>(input);
+    // check if out of bound
+    // std::isinf is used for check long double bounds
+    if (std::isinf(sum)) {
+        throw std::runtime_error("Overflow occured for sum during calculating statistic metrics! Check your data!");
+    }
+    return;
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::BasicStatsTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::BasicStatsTrainingOnlyPolicy<T>::BasicStatsTrainingOnlyPolicy(void) :
+    _count(0),
+    _first_element_flag(true) {
+}
+
+
+
+template <typename T>
+void Details::BasicStatsTrainingOnlyPolicy<T>::fit(InputType const &input) {
+    if(Microsoft::Featurizer::Traits<T>::IsNull(input))
+        return;
+    update_basic_statistics(input, _min, _max, _count, _first_element_flag);
+
+}
+
+template <typename T>
+BasicStatisticalAnnotationData<T> Details::BasicStatsTrainingOnlyPolicy<T>::complete_training(void) {
+    if (_count != 0) {
+        assert(_min<=_max);
+    }
+    return BasicStatisticalAnnotationData<T>(std::move(_min), std::move(_max), std::move(_count));
+
+}
+
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::StandardStatsTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+template <typename T>
+Details::StandardStatsTrainingOnlyPolicy<T>::StandardStatsTrainingOnlyPolicy(void) :
+    _sum(0),
+    _average(0) {
+        BasicStatsTrainingOnlyPolicy<T>::_min = 0;

```

Comment:
Are these statements necessary? Won't initialization be automatically handled in the first call to update_basic_statistics.